---
title: "Introduction to PhysioSpaceMethods"
author: "A. H. Esfahani"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to PhysioSpace}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

PhysioSpace is a robust statistical method for relating high dimensional omics
data sets^[Lenz, Michael, et al. "PhysioSpace: relating gene expression
experiments from heterogeneous sources using shared physiological processes."
PLoS One 8.10 (2013): e77627]. It is designed to take advantage of the vast
availability of public omics data, which in combination with statistical
approaches makes a potent tool capable of analyzing heterogeneous biological
data sets.

PhysioSpaceMethods is a R package which provides an implementation of
PhysioSpace method alongside other handy functions for making PhysioSpace an
easily accessible tool for R users.




### Table of Contents
**[Installation Instructions](#installation-instructions)**<br>
**[Usage Instructions](#usage-instructions)**<br>
**[Session Info](#sessioninfo)**<br>

## Installation Instructions
### Installing via Devtools (Recommended method):
Easiest way to install PhysioSpaceMethods is via <a
href="https://cran.r-project.org/web/packages/devtools/">devtools</a> package.
After installing devtools from cran, you can install PhysioSpaceMethods by:
```r
devtools::install_github(repo = "JRC-COMBINE/PhysioSpaceMethods",
build_vignettes = TRUE)
```

### Alternative installation methods (Manual download):
In case you encountered any problem while installing PhysioSpaceMethods, you
can download the repository directly and 
install the package from downloaded local files.
In your terminal, first clone the repository in your desired repository:
```Shell
cd [Your desired directory]
git clone https://github.com/JRC-COMBINE/PhysioSpaceMethods.git
```
Then install the downloaded package using <a
href="https://cran.r-project.org/web/packages/devtools/">Devtools</a>:
```Shell
R -e "devtools::install_local('./PhysioSpaceMethods/', build_vignettes =
TRUE)"
```

## Usage Instructions
PhysioSpaceMethods can map user samples inside a physiological space,
calculated beforehand from a compendium of known samples. Few examples are
provided here to demonstrate how to use this package.

### Example One: E-MTAB-2836 Analysis
With the first example we will show how PhysioSpace can relate RNA-seq data to
data generated with micro-array.
The data set used in this example is <a
href="https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2836/">
E-MTAB-2836</a>, a RNA-seq atlas of coding RNA from tissue samples of 122
_**human**_ individuals representing 32 
different tissues, stored on ebi's <a href="https://www.ebi.ac.uk/gxa/home">
Expression Atlas</a>.

To start the analysis, first we prepare the E-MTAB-2836 for our pipeline. You
can download the data set manually from 
<a href="https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2836/Downloads">this
page (the 'Summary of the expression results for this experiment ready to view
in R' link)</a>, or use the following command in R:
```r
#Download:
download.file(url = "https://www.ebi.ac.uk/gxa/experiments-content/E-MTAB-2836/static/E-MTAB-2836-atlasExperimentSummary.Rdata",
              destfile= "E-MTAB-2836-atlasExperimentSummary.Rdata") # We're
              #downloading into the working directory, obviously using any
              #other directory is possible.
```

After downloading (and normalising if necessary), the data can be analysed
using "calculatePhysioMap()" function. It is important to note that
'InputData' of this function has specific format requirements which are needed
to be met for the function to perform properly. Detailed description of input
requirements can be found in calculatePhysioMap's help page. In short,

1. In case InputData is a **matrix**, it should be:
    a. a matrix,
    b. with genes in rows, identified by Entrez IDs which are stored in 
    'rownames' of the matrix. And
    c. samples in columns, identified by sample names stored in 'colnames' 
    of the matrix. Lastly, 
    d. values of InputData matrix should be **relative**, e.g. fold changes of 
    genes, rather than pure gene expressions. 
    
2. In case InputData is a **SummarizedExperiment object**, it is supposed to
have:
    a. a component named 'EntrezID' in its rowData, containing Entrez IDs of 
    rows in the object. And
    b. a component named 'SampleName' in its colData, containing name or the
    main annotation of the samples in the object. Lastly,
    c. the assay in the SummarizedExperiment object (accessible by using 
    SummarizedExperiment::assay(obj)) has to have **relative** values, e.g. 
    fold changes of genes, rather than pure gene expressions.
    
3. In case user has their own list of significantly up and down regulated 
genes, it is also possible for InputData to be a list: 
    a. containing Entrez IDs (or any other identifier which is used as 
    rownames in 'Space') of up regulated genes in InputData[[1]], and 
    b. Entrez IDs (or any other identifier which is used as rownames 
    in 'Space') of down regulated genes in InputData[[2]].

For demonstration purposes, in this example we prepare the InputData as 
a matrix, and as a SummarizedExperiment.

#### E-MTAB-2836 InputData as a matrix:
We prepare E-MTAB-2836 for calculatePhysioMap() in four steps:

* **Converting gene expression data into a matrix:** In this example 
E-MTAB-2836-atlasExperimentSummary.Rdata contains a RangedSummarizedExperiment
object, from which we have to extract a gene expression matrix:
```r
#Making the gene expression matrix:
library(SummarizedExperiment) #SummarizedExperiment is needed for working with
                              #RangedSummarizedExperiment objects.
load("E-MTAB-2836-atlasExperimentSummary.Rdata") #Loading the object into R
EMTAB2836CountMatrix <- assay(experimentSummary$rnaseq)
```

* **Having genes in rows and samples in columns, with Entrez IDs in
rownames:** After converting our gene expression data 
to a matrix, we have to make sure that we have genes in rows and samples in
columns, with genes identified by Entrez IDs stored in 'rownames' of the
matrix. In our case we already have genes as rows and samples as columns in
EMTAB2836CountMatrix. However, Ensembl IDs are stored in rownames, which have
to be converted to Entrez IDs. There are numerous ways for this conversion.
For example by using the 
<a href="https://david.ncifcrf.gov/conversion.jsp">David's Gene ID Conversion
Tool</a>, or by using 
Ensembl's <a href="https://www.ensembl.org/biomart/">BioMart</a> or <a
href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">
biomaRt's package</a> in R.
Here we used biomaRt's package:
```r
#Converting Ensembl to Entrez IDs:
library(biomaRt)
humaRt <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                                  dataset = "hsapiens_gene_ensembl")
ConvTabelle <- getBM(attributes = c("ensembl_gene_id","entrezgene"),
                     filters = "ensembl_gene_id", 
                     values = rownames(EMTAB2836CountMatrix),
                     mart = humaRt)
rownames(EMTAB2836CountMatrix) <-
              ConvTabelle$entrezgene[match(rownames(EMTAB2836CountMatrix),
              ConvTabelle$ensembl_gene_id)]
EMTAB2836CountMatrix <-
          EMTAB2836CountMatrix[!is.na(rownames(EMTAB2836CountMatrix)),] 
          #We remove the IDs that couldn't be converted and turned into NAs.
```

* **Writing sample names in colnames:** This step is not necessary for proper
calculation, but it is highly recommended as it helps in proper inference of
the result matrix. Sample names should be short, and to the point. Since in
this example we aim to match RNA-seq counts of different human tissues to 
their corresponding tissues in a micro-array compendium, we use the tissue 
names as colnames of EMTAB2836CountMatrix:
```r
#Assigning colnames:
colnames(EMTAB2836CountMatrix) <-
                            colData(experimentSummary$rnaseq)$organism_part
```

* **Having _RELATIVE_ values for gene expression:** PhysioSpace method expects
relative values as input, i.e. it assumes the most 
positive values correspond to up-regulated genes and most negative values
correspond to down-regulated genes. Easiest way to 
calculate this relative value is by calculating fold change, although more
proper statistical tests could result in better performance.
In this specific example we use fold change:
```r
#Calculating Fold-Changes:
EMTAB2836CountMatrixRelativ <- EMTAB2836CountMatrix -
                                        apply(EMTAB2836CountMatrix,1,mean)
```
We used the gene-wise mean value of the whole data set as a virtual control
sample and calculated the fold changes based on this
virtual control, since all data points in E-MTAB-2836 are biopsy samples and
there are no actual control samples. At the same time, because of the high
number of samples in E-MTAB-2836, the mean value is a good measure of
background noise on each gene. Therefore,
mean values work great as controls to compare against.
As mentioned above, there are more sophisticated ways for this calculation,
one example could be to use the signed p value of a statistical test in
logarithm scale (which will come later for the other example
in this vignette).

#### E-MTAB-2836 InputData as a SummarizedExperiment object:
As an alternative, InputData could be a SummarizedExperiment object when
passed to the calculatePhysioMap() function. Here, we prepare the object
in three steps:

* **Setting up 'EntrezID' component in rowData:** After loading E-MTAB-2836
into working environment, we have to add a component named 'EntrezID' to
row data of the loaded object. This new added component should contain
Entrez IDs of the rows of data in the SummarizedExperiment object:
```r
#Loading the object into R:
suppressPackageStartupMessages(library(SummarizedExperiment))
load("E-MTAB-2836-atlasExperimentSummary.Rdata")
EMTAB2836_SEObj <- experimentSummary$rnaseq
#
#Converting Ensembl to Entrez IDs, and making EntrezID in rowData:
library(biomaRt)
humaRt <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                                  dataset = "hsapiens_gene_ensembl")
ConvTabelle <- getBM(attributes = c("ensembl_gene_id","entrezgene"),
                     filters = "ensembl_gene_id", 
                     values = rownames(EMTAB2836_SEObj),
                     mart = humaRt)
rowData(EMTAB2836_SEObj) <- data.frame("EntrezID" = 
              ConvTabelle$entrezgene[match(rownames(EMTAB2836_SEObj),
              ConvTabelle$ensembl_gene_id)])
EMTAB2836_SEObj <-
          EMTAB2836_SEObj[!is.na(rowData(EMTAB2836_SEObj)$EntrezID),] 
          #We remove the rows with Entrez IDs that couldn't be 
          #converted and turned into NAs.
```
We used <a 
href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">
biomaRt's package</a> for converting Ensembl to Entrez IDs.

* **Setting up 'SampleName' component in colData:** This step is not 
necessary for proper calculation, but it is highly recommended as it 
helps in proper inference of the result matrix. We have to assign a 
component named 'SampleName' to col data of the object, which contains
the main annotation of samples in the data set under analysis. 
In this example, we aim to detect the tissue type of samples in 
E-MTAB-2836. So our target annotation is the source tissue name of each
sample, already available as 'organism_part' component of col data.
We just have to rename 'organism_part' to 'SampleName':
```r
#Assigning SampleName:
names(colData(EMTAB2836_SEObj))[names(colData(EMTAB2836_SEObj)) == 
                                    "organism_part"] <- "SampleName"
```

* **Having _RELATIVE_ values in Assay:** PhysioSpace method expects
relative values is assay of SummarizedExperiment input object, i.e. 
it assumes the most positive values correspond to up-regulated genes 
and most negative values correspond to down-regulated genes. 
As mentioned above, in this example we use universal mean as control 
and calculate fold changes based on that reference:
```r
#Calculating Fold-Changes:
assay(EMTAB2836_SEObj) <- assay(EMTAB2836_SEObj) - 
                                        apply(assay(EMTAB2836_SEObj),1,mean)
```

Having the proper format for the input, the main calculation can be done
easily by the function calculatePhysioMap.

calculatePhysioMap() has two required input arguments: InputData, which is the
relative gene expression matrix (or SummarizedExperiment obj.), and Space, 
which is the Physiological Space in which we want to map our input data. In
this example we use 
[LUKK](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2974261/) space from
[HumanPhysioSpace](https://github.com/JRC-COMBINE/HumanPhysioSpace) package:
```r
#Main calculation:
library(PhysioSpaceMethods)
library(HumanPhysioSpace) # you can install this package from
                          #https://github.com/JRC-COMBINE/HumanPhysioSpace
RESULTS <- calculatePhysioMap(InputData = EMTAB2836CountMatrixRelativ, 
                                                    Space = HS_LUKK_Space)
#Main calculation, for SummarizedExperiment input obj.:
RESULTS_SE <- calculatePhysioMap(InputData = EMTAB2836_SEObj, 
                                                    Space = HS_LUKK_Space)
```
For more information about the available Spaces in HumanPhysioSpace package,
detail explanation about
HS_LUKK_Space and information about other input options of
calculatePhysioMap(), we recommend the reader to check the documentation
of this package and HumanPhysioSpace.

```r
#Check to see if the results are the same, regardless of input format:
identical(RESULTS, RESULTS_SE)
```
As expected, results are the same for both input types of matrix and
SummarizedExperiment obj.. In this vignette from here onward, we do 
all the calculations in the matrix format.

In cases with large number of input samples, we recommend running
calculatePhysioMap() in parallel:
```r
#Main calculation in parallel:
RESULTS <- calculatePhysioMap(InputData = EMTAB2836CountMatrixRelativ, 
                      Space = HS_LUKK_Space, PARALLEL = T, NumbrOfCores = 4)
```

The output of calculatePhysioMap(), which here we called 'RESULTS', is a
matrix with the same number of columns as the number of 
samples in 'InputData', and the same number of rows as the number of axes
(Columns) in the 'Space'. The 
value in row M and Column N in RESULTS is the mapped values of N<sup>th</sup>
sample on M<sup>th</sup> axis of the Space.

In our example there are 200 samples. We randomly choose 5 samples out of 200
and show the matching between RNA-seq input 
data set to micro-array reference compendium is successful:
```r
#Choosing 5 random samples:
set.seed(seed = 0) #So results would be reproducible 
RESULTS5Random <- RESULTS[,sample(x = 1:ncol(RESULTS), size = 5)]
#Plotting the results:
PhysioHeatmap(PhysioResults = RESULTS5Random, main = "RNA-seq vs Microarray",
            SymmetricColoring = T, SpaceClustering = T, Space = HS_LUKK_Space,
                                                          ReducedPlotting = 5)
```
![Figure 1. Similarities heat map of RNA-seq 
tissue samples of E-MTAB-2836 vs. Lukk et al. Space made from 
micro-array tissue samples](RESULTSHeatmap.png "RESULTS Heatmap")

As shown in Fig 1, we expect to have the highest values (most red) in the
intersection of
each column with its corresponding tissue in rows. From the 5 samples we
analysed, "skeletal muscle tissue", "esophagus" and "placenta" 
are clearly matched to their corresponding tissues from micro-array space.
From two remaining samples, "vermiform appendix" is matched to blood,
and that is because there is no appendix tissue sample in Lukk data set.
Considering that, matching to blood makes sense because the vermiform appendix 
biopsy is very likely to contain a large portion of blood, hence the
conversion from RNA-seq to micro-array is successful in this sample
as well. Same is true for the sample "smooth muscle tissue": there are many
organs from which this smooth muscle sample could be acquired, and since no
more extra information is provided in E-MTAB-2836 about this sample except
that the sample is smooth muscle tissue from a female adult human, based on
our results it is highly probable that the smooth muscle sample is acquired
from uterus.

### Example Two: GSE106635
To show the application of PhysioSpace in plant stress analysis, we will
analyse <a 
href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE106635">
GSE106635</a> from NCBI's <a href="https://www.ncbi.nlm.nih.gov/geo/">
Gene Expression Omnibus or GEO</a> as our second example.

There are numerous ways to acquire a data set from GEO, for example by using
<a href="https://bioconductor.org/packages/release/bioc/html/GEOquery.html">
GEOquery</a>. Here we directly download 
the CEL files and normalise them with <a
href="https://www.bioconductor.org/packages/release/bioc/html/affy.html">
affy</a> package:
```r
#Download and untar:
download.file(url = 
    "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE106635&format=file",
              destfile= "GSE106635_RAW.tar") # We're downloading into the
              #working directory, obviously using any other directory is
              #possible.
untar(tarfile = "GSE106635_RAW.tar", exdir = "GSE106635_RAW")

#Normalising:
library(affy)
GSE106635ESet <-
  justRMA(
    filenames = list.files(path = "GSE106635_RAW", full.names = T)
  )
```

After downloading and normalising, we need to prepare the data before using 
it as InputData of calculatePhysioMap function. 
In this example we use signed p values as relative values in InputData:

* **Convert gene expression data into a matrix of signed p values:** we
calculate the signed p value of Student's t-test using the 
<a href="https://bioconductor.org/packages/release/bioc/html/limma.html">
limma</a> package (for more information about linear 
modelling and statistical testing, you can check limma package
documentation).
```r
#Converting expression set to matrix:
GSE106635ESetExp <- exprs(GSE106635ESet)
#Modelling and statistical testing:
library(limma)
DESIGN <- model.matrix(object = ~0+annot, data =
data.frame(annot=c(rep("WT_Ctrl",2),rep("Mut_Ctrl",2),
                      rep("WT_Cold",2),rep("Mut_Cold",2))))
FITMain <- lmFit(GSE106635ESetExp, DESIGN)
cont.matrix <- makeContrasts(contrasts =
c("annotWT_Cold-annotWT_Ctrl","annotMut_Cold-annotMut_Ctrl"), levels=DESIGN)
FITCont <- contrasts.fit(FITMain, cont.matrix)
FITCont <- eBayes(FITCont)
WTResults <- topTable(fit = FITCont, coef = "annotWT_Cold-annotWT_Ctrl",
                      adjust.method="BH", number = Inf, sort.by = "none")
MutResults <- topTable(fit = FITCont, coef = "annotMut_Cold-annotMut_Ctrl",
                        adjust.method="BH", number = Inf, sort.by = "none")
GSE106635SignedPValues <-
cbind(-log2(WTResults$adj.P.Val)*sign(WTResults$logFC),
      -log2(MutResults$adj.P.Val)*sign(MutResults$logFC))
rownames(GSE106635SignedPValues) <- rownames(GSE106635ESetExp)
```

* **Have genes in rows and samples in columns, with Entrez IDs in rownames:**
For converting the Affymetrix ATH1 probe IDs to Entrez IDs
we use 
<a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">
biomaRt's package</a>:
```r
#Converting AffyID to EntrezID:
library(biomaRt)
planaRt <- useMart(biomart = "plants_mart", host = "plants.ensembl.org",
                                            dataset = "athaliana_eg_gene")
ConvTabelle <- getBM(attributes = c("affy_ath1_121501","entrezgene"),
                            filters = "affy_ath1_121501", 
                              values = rownames(GSE106635SignedPValues),
                                  mart = planaRt)
rownames(GSE106635SignedPValues) <-
ConvTabelle$entrezgene[match(rownames(GSE106635SignedPValues),
                                  ConvTabelle$affy_ath1_121501)]
GSE106635SignedPValues <-
          GSE106635SignedPValues[!is.na(rownames(GSE106635SignedPValues)),]
```

* **Have RELATIVE values for gene expression:** In this example, since we
calculated signed p values, numbers in GSE106635SignedPValues are already
relative.

* **Writing sample names in colnames:** In GSE106635SignedPValues, first
column corresponds to wild type Col-0 response to cold stress, and second
column to TabZIP6-overexpressed Arabidopsis line (L20) response to cold
stress. We will name samples as follows:
```r
#Writing names into colnames:
colnames(GSE106635SignedPValues) <- c("WT_Cold","Mut_Cold")
```

Now that we prepared the input, the main calculation can be done using the
function calculatePhysioMap():
```r
#Main calculation:
library(PhysioSpaceMethods)
library(PlantPhysioSpace)
RESULTS <- calculatePhysioMap(InputData = GSE106635SignedPValues, 
                                              Space = AT_Stress_Space)
```
Note that as mentioned in the last example, calculatePhysioMap() has to have
at least two inputs: 'InputData' which in this case is the signed p value
matrix, and 'Space' which is the Physiology Space in which we want the
InputData to be mapped. In this example we used AT_Stress_Space from 
<a href="https://github.com/JRC-COMBINE/PlantPhysioSpace">
PlantPhysioSpace</a> package as Space. For more information about the
available plant spaces and detailed explanation about AT_Stress_Space, we
recommend the reader to check the documentation of the 
<a href="https://github.com/JRC-COMBINE/PlantPhysioSpace">
PlantPhysioSpace</a> package.

All samples in GSE106635 are under cold stress condition, so we expect to see
high values (similarities) on the 'Cold' dimension of 
the AT_Stress_Space:
```r
#Plotting the results:
PhysioHeatmap(PhysioResults = RESULTS, main = "Stress Analysis of GSE106635",
SymmetricColoring = T, SpaceClustering = T, Space = AT_Stress_Space, 
                                                              PlotSize = 21)
```
![Figure 2. Similarities heat map of cold stressed A.thaliana samples 
of GSE106635 vs. AT_Stress_Space](RESULTSGSE106635.png "RESULTS Heatmap")

As shown in Fig. 2, samples of GSE106635 are detected as plants under cold
stress. Looking deeper into the RESULTS, analysis points toward a higher cold
stress response in wild type Arabidopsis, in comparison to
TabZIP6-overexpressed ecotype (RESULTS[4,1] > RESULTS[4,2]). This conclusion
is in line with the conclusion published by the cited paper of GSE106635^[Cai
W, Yang Y, Wang W, Guo G et al. Overexpression of a wheat (Triticum aestivum
L.) bZIP transcription factor gene, TabZIP6, decreased the freezing tolerance
of transgenic Arabidopsis seedlings by down-regulating the expression of CBFs.
Plant Physiol Biochem 2018 Mar;124:100-111.].


## SessionInfo()
####
R version 3.5.1 (2018-07-02) 

Platform: x86_64-apple-darwin15.6.0 (64-bit)

Running under: macOS Sierra 10.12.6

####
Matrix products: default

BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib

LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

####
locale:

[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

####
attached base packages:

[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base

####
other attached packages:

 [1] PlantPhysioSpace_0.9.0      limma_3.37.3               
 [3] ath1121501cdf_2.18.0        affy_1.59.0                
 [5] HumanPhysioSpace_0.0.0.9000 biomaRt_2.37.3             
 [7] SummarizedExperiment_1.11.6 DelayedArray_0.7.24        
 [9] BiocParallel_1.15.8         matrixStats_0.54.0         
[11] Biobase_2.41.2              GenomicRanges_1.33.13      
[13] GenomeInfoDb_1.17.1         IRanges_2.15.16            
[15] S4Vectors_0.19.19           BiocGenerics_0.27.1        
[17] PhysioSpaceMethods_0.99.1  


####
loaded via a namespace (and not attached):

 [1] minqa_1.2.4            colorspace_1.3-2       class_7.3-14          
  [4] rprojroot_1.3-2        htmlTable_1.12         XVector_0.21.3        
  [7] base64enc_0.1-3        rstudioapi_0.7         mice_3.3.0            
 [10] affyio_1.51.0          bit64_0.9-7            mvtnorm_1.0-8         
 [13] AnnotationDbi_1.43.1   missMDA_1.13           codetools_0.2-15      
 [16] splines_3.5.1          leaps_3.0              doParallel_1.0.11     
 [19] geneplotter_1.59.0     knitr_1.20             DMwR_0.4.1            
 [22] Formula_1.2-3          nloptr_1.0.4           broom_0.5.0           
 [25] annotate_1.59.1        cluster_2.0.7-1        httr_1.3.1            
 [28] compiler_3.5.1         backports_1.1.2        assertthat_0.2.0      
 [31] Matrix_1.2-14          lazyeval_0.2.1         acepack_1.4.1         
 [34] htmltools_0.3.6        prettyunits_1.0.2      tools_3.5.1           
 [37] bindrcpp_0.2.2         gtable_0.2.0           glue_1.3.0            
 [40] GenomeInfoDbData_1.1.0 dplyr_0.7.6            FactoMineR_1.41       
 [43] Rcpp_0.12.18           preprocessCore_1.43.0  gdata_2.18.0          
 [46] nlme_3.1-137           iterators_1.0.10       stringr_1.3.1         
 [49] lme4_1.1-17            gtools_3.8.1           XML_3.98-1.13         
 [52] pan_1.6                zlibbioc_1.27.0        MASS_7.3-50           
 [55] zoo_1.8-3              scales_0.5.0           BiocInstaller_1.31.1  
 [58] hms_0.4.2              RColorBrewer_1.1-2     yaml_2.2.0            
 [61] quantmod_0.4-13        curl_3.2               memoise_1.1.0         
 [64] gridExtra_2.3          ggplot2_3.0.0          rpart_4.1-13          
 [67] latticeExtra_0.6-28    stringi_1.2.4          RSQLite_2.1.1         
 [70] genefilter_1.63.0      foreach_1.4.4          checkmate_1.8.5       
 [73] TTR_0.23-3             caTools_1.17.1.1       rlang_0.2.1           
 [76] pkgconfig_2.0.1        bitops_1.0-6           evaluate_0.11         
 [79] lattice_0.20-35        ROCR_1.0-7             purrr_0.2.5           
 [82] bindr_0.1.1            htmlwidgets_1.2        bit_1.1-14            
 [85] tidyselect_0.2.4       plyr_1.8.4             magrittr_1.5          
 [88] DESeq2_1.21.12         R6_2.2.2               gplots_3.0.1          
 [91] Hmisc_4.1-1            mitml_0.3-6            DBI_1.0.0             
 [94] pillar_1.3.0           foreign_0.8-71         xts_0.11-0            
 [97] survival_2.42-6        scatterplot3d_0.3-41   abind_1.4-5           
[100] RCurl_1.95-4.11        nnet_7.3-12            tibble_1.4.2          
[103] crayon_1.3.4           jomo_2.6-3             KernSmooth_2.23-15    
[106] rmarkdown_1.10         progress_1.2.0         locfit_1.5-9.1        
[109] grid_3.5.1             data.table_1.11.4      blob_1.1.1            
[112] digest_0.6.15          flashClust_1.01-2      xtable_1.8-2          
[115] tidyr_0.8.1            munsell_0.5.0         

